sycamore/transforms/extract_table.py

```diff
@@ -108,3 +114,152 @@ def extract_tables(self, document: Document) -> Document:
+
+    Cached table is in Json format as below:
+    {
+      "tables": [
+        {
+          "page_number": xx,
+          "title": "xxxx",
+          "content": "xxx"
+          "bbox": [xxx, xxx, xxx, xxx],
+          "footers": ["xxx", "xxx"]
+        },
+        ...
+      ]
+    }
+    bbox format is in [left/width, top/height, right/width, bottom/height]
+    """
+
+    def __init__(
+        self,
+        s3_cache_location,
+        s3_textract_upload_path: Optional[str] = None,
+        profile_name: Optional[str] = None,
+        region_name: Optional[str] = None,
+        kms_key_id: str = "",
+    ):
+        self._s3_cache_location = s3_cache_location
+        self._s3_textract_upload_path = s3_textract_upload_path
+        self._profile_name = profile_name
+        self._region_name = region_name
+        self._kms_key_id = kms_key_id
+        self._s3 = boto3.client("s3")
+
+    @staticmethod
+    def _parse(json_data):
+        """
+        Parse json format cache into a list of table elements
+        """
+        tables = []
+        for table in json_data["tables"]:
+            element = Element()
+            element.type = "Table"
+            properties = element.properties
+            properties["page_number"] = table["page_number"]
+            element.properties = properties
+
+            text_list = []
+            if table.get("title"):
+                text_list.append(table["title"])
+            text_list.append(table["content"])
+            element.bbox = BoundingBox(*table["bbox"])
+            text_list.extend(table.get("footers", []))
+
+            element.text_representation = "\n".join(text_list)
+            tables.append(element)
+        return tables
+
+    def _get_table(self, cache_id) -> Optional[list[Element]]:
+        """Get cache from S3"""
+        try:
+            parts = self._s3_cache_location.replace("s3://", "").split("/", 1)
+            bucket = parts[0]
+            key = "/".join([parts[1], cache_id]) if len(parts) == 2 else cache_id
+            response = self._s3.get_object(Bucket=bucket, Key=key)
+        except ClientError as e:
+            if e.response["Error"]["Code"] == "NoSuchKey":
+                return None
+            else:
+                raise
+
+        tables = self._parse(json.loads(response["Body"].read()))
+        return tables
+
+    def _put_table(self, cache_id: str, table: dict):
+        """Put table into S3"""
+        parts = self._s3_cache_location.replace("s3://", "").split("/", 1)
+        bucket = parts[0]
+        key = "/".join([parts[1], cache_id]) if len(parts) == 2 else cache_id
+        json_str = json.dumps(table, indent=2)
+        self._s3.put_object(Body=json_str, Bucket=bucket, Key=key)
+
+    def extract_tables(self, document: Document) -> Document:
+        # a list holding 1 based page number
+        table_pages = [element.properties["page_number"] for element in document.elements if element.type == "Table"]
```

Comment:
Do we want to have a option/fallback to the current behavior where we call for the entire doc? People who don't/can't run the DETR model (e.g. on the mac) might not have annotated labels, and caching would still be useful.

Reply:
good point, would expose a knob for this.